### PR TITLE
Multiselect Lookup (LWC): Performance Improvement on large Data (a2R1T000002QLM5UAO)

### PR DESCRIPTION
**Pull request by rajas94**
Please look into the usage of lookup [here](https://github.com/Up2Go/cprm/blob/3f7743eef0a83d49405838c56e1f0d1a1e475e46/sfdx-source/UP2GO_CPRM_BM/main/default/aura/attributeEditor/attributeEditor.cmp#L26) in order to get an idea of lookup properties. Please feel free to reach out to me if you have any ideas/questions.

### CODE REVIEW

File: src/main/default/lwc/lookup/lookup.html [hunk 1]
```diff
@@ -146,13 +148,13 @@
             </template>
             <!-- Multi-selection end -->
 
-            <!-- Errors start -->
-            <template for:each={errors} for:item="error">
-                <label key={error.id} role="alert" class="slds-form-element__label slds-var-m-top_xx-small form-error"
-                    >{error.message}</label
-                >
+            <!-- Error start -->
```
Comment (gorazd-up2go) [inline src/main/default/lwc/lookup/lookup.html]: I would not change how error is displayed. Before you could pass in array and I would not change that because it can happen we have more errors.

File: src/main/default/lwc/lookup/lookup.js [hunk 2]
```diff
@@ -112,7 +118,8 @@ export default class Lookup extends LightningElement {
 
         // Compare clean new search term with current one and abort if identical
         const newCleanSearchTerm = newSearchTerm.trim().replace(/\*/g, '').toLowerCase();
-        if (this._cleanSearchTerm === newCleanSearchTerm) {
+
+        if (!this.dispatchSearchEventOnEachRequest && this._cleanSearchTerm === newCleanSearchTerm) {
```
Comment (gorazd-up2go) [inline src/main/default/lwc/lookup/lookup.js]: I think this logic regarding dispatchSearchEventOnEachRequest should not be implemented in this component but on upper componenent that is using it. What is purpose of this? To avoid to many events?

File: src/main/default/lwc/lookup/lookup.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { LightningElement, api } from 'lwc';
 
-const MINIMAL_SEARCH_TERM_LENGTH = 2; // Min number of chars required to search
-const SEARCH_DELAY = 300; // Wait 300 ms after user stops typing then, peform search
+const MINIMAL_SEARCH_TERM_LENGTH = 0; // Min number of chars required to search
```
Comment (gorazd-up2go) [inline src/main/default/lwc/lookup/lookup.js]: A question here is that on some components it is good to start it from 0. Maybe this could be set over api and leave default on 0. And api woule overwrite it.

**Review by gorazd-up2go — CHANGES_REQUESTED** [review]
lookupRendering.test.js is missing tests for new functionality for example to disable and showing label on hasMoreResultsToDisplay

File: src/main/default/lwc/lookup/lookup.js [hunk 4]
```diff
@@ -14,9 +13,12 @@ export default class Lookup extends LightningElement {
     @api isMultiEntry = false;
     @api hideResults = false;
     get showResults() {return this.isMultiEntry && !this.hideResults}
-    @api errors = [];
     @api scrollAfterNItems;
     @api iconResources = {};
+    @api errors;
```
Comment (gorazd-up2go) [inline src/main/default/lwc/lookup/lookup.js]: I would use original empty array:
@api errors = [];

File: src/main/default/lwc/lookup/lookup.js [hunk 5]
```diff
@@ -14,9 +13,12 @@ export default class Lookup extends LightningElement {
...
+    @api hasError;
```
Comment (gorazd-up2go) [inline src/main/default/lwc/lookup/lookup.js]: I think this is unnecessary. The same can be achieved with a getter:

get hasErrors() {
    return this.errors.length !== 0;
}

**Comment by gorazd-up2go** [comment]
@rajas94  There is one more minor improvement I think of and somebody else should look at it as well.

**Review by gorazd-up2go — APPROVED** [review]
I think it is ok but somebody else should look at it.

File: src/main/default/lwc/lookup/__tests__/lookupEventFiring.test.js [hunk 6]
```diff
@@ -57,7 +57,7 @@ describe('c-lookup event fires', () => {
         });
     });
 
-    it('search event does not fires when search term is too short', () => {
+    it('fire search event even search term is empty', () => {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/__tests__/lookupEventFiring.test.js]: this seems to be a special requirement for your story. I don't think this should apply to all usages. Did you think about encapsulating your changes in a wrapping component?

File: src/main/default/lwc/lookup/__tests__/lookupRendering.test.js [hunk 7]
```diff
@@ -153,15 +153,28 @@ describe('c-lookup rendering', () => {
...
+    it('show refine search text when passed search results are less than actual search results', () => {
+        // Create element
+        const element = createElement('c-lookup', {
+            is: Lookup
+        });
+        element.setSearchResults(SAMPLE_SELECTION_ITEMS, 5);
+        document.body.appendChild(element);
+
+        // Verify errors
+        const refineSearch = element.shadowRoot.querySelector('span.slds-float_right');
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/__tests__/lookupRendering.test.js]: not sure if span.slds-float_right is a good selector.
refineSearchErrorMessage

File: src/main/default/lwc/lookup/lookup.html [hunk 8]
```diff
@@ -146,13 +148,15 @@
...
+            <template if:true={hasErrors}>
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.html]: this if should not be necessary, looping through an empty list results in the same output.
the best would be to revert this change so we are not spreadingt to far away from the original branch

File: src/main/default/lwc/lookup/lookup.html [hunk 9]
```diff
@@ -1,5 +1,6 @@
 <template>
     <div class="slds-form-element">
+        <span if:true={hasMoreResultsToDisplay} class="slds-float_right slds-text-body_small">Please refine your search to see more results</span>
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.html:3]: why is this message not handeled like all the other errors?

File: src/main/default/lwc/lookup/lookup.js [hunk 10]
```diff
@@ -14,9 +13,11 @@ export default class Lookup extends LightningElement {
     @api isMultiEntry = false;
     @api hideResults = false;
     get showResults() {return this.isMultiEntry && !this.hideResults}
-    @api errors = [];
     @api scrollAfterNItems;
     @api iconResources = {};
+    @api errors = [];
+    @api disabled;
+    @api minimalSearchTermLength = 0;
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: the default should stay at 2

Comment (itsmebasti) [reply src/main/default/lwc/lookup/lookup.js]: how about 
@api minimalSearchTermLength = MINIMAL_SEARCH_TERM_LENGTH;
to not change so many things in this component, but make it more extending original code instead of replacing it,
this will be much easier to merge and maintain

File: src/main/default/lwc/lookup/lookup.js [hunk 11]
```diff
@@ -47,7 +49,9 @@ export default class Lookup extends LightningElement {
     }
 
     @api
-    setSearchResults(results) {
+    setSearchResults(results, totalSearchResults) {
+        this._totalSearchResults = totalSearchResults;
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: this information should not be passed in again aand again, since it will alway be the same per element.
pass it in via @api attributes and than extend the logic araound it using this infomration

File: src/main/default/lwc/lookup/lookup.js [hunk 12]
```diff
@@ -112,16 +116,13 @@ export default class Lookup extends LightningElement {
 
         // Compare clean new search term with current one and abort if identical
         const newCleanSearchTerm = newSearchTerm.trim().replace(/\*/g, '').toLowerCase();
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js:122]: do not remove this. instead pass in an api boolean, wether you want to execute the search for empty values too, so the default behaviour does not change

File: src/main/default/lwc/lookup/lookup.js [hunk 13]
```diff
@@ -184,7 +187,7 @@ export default class Lookup extends LightningElement {
 
     handleInput(event) {
         // Prevent action if selection is not allowed
-        if (!this.isSelectionAllowed()) {
+        if (!this.isSelectionAllowed() || this.disabled) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: how can someone input something to a disabled field?

File: src/main/default/lwc/lookup/lookup.js [hunk 14]
```diff
@@ -246,13 +249,14 @@ export default class Lookup extends LightningElement {
         this.template.querySelector('input').focus();
     }
 
-    handleFocus() {
+    handleFocus(evt) {
         // Prevent action if selection is not allowed
-        if (!this.isSelectionAllowed()) {
+        if (!this.isSelectionAllowed() || this._hasFocus) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: how can something befocused, when it is already focused?

File: src/main/default/lwc/lookup/lookup.js [hunk 15]
```diff
@@ -281,27 +296,26 @@ export default class Lookup extends LightningElement {
 
     get getContainerClass() {
         let css = 'slds-combobox_container slds-has-inline-listbox ';
-        if (this._hasFocus && this.hasResults()) {
+        if (this._hasFocus && this.loading) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: why did you change this?

File: src/main/default/lwc/lookup/lookup.js [hunk 16]
```diff
@@ -281,27 +296,26 @@ export default class Lookup extends LightningElement {
...
             css += 'has-custom-error';
         }
         return css;
     }
 
     get getDropdownClass() {
         let css = 'slds-combobox slds-dropdown-trigger slds-dropdown-trigger_click ';
-        const isSearchTermValid = this._cleanSearchTerm && this._cleanSearchTerm.length >= MINIMAL_SEARCH_TERM_LENGTH;
-        if (this._hasFocus && this.isSelectionAllowed() && (isSearchTermValid || this.hasResults())) {
+        if (this._hasFocus && this.isSelectionAllowed()) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: not sure if we should change this

File: src/main/default/lwc/lookup/lookup.js [hunk 17]
```diff
@@ -374,4 +388,12 @@ export default class Lookup extends LightningElement {
     get isExpanded() {
         return this.hasResults();
     }
+
+    get hasMoreResultsToDisplay() {
+        return this.searchResultsLocalState.length !== 0 && this.searchResultsLocalState.length < this._totalSearchResults;
+    }
+
+    get hasErrors() {
+        return this.errors && this.errors.length > 0;
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: maybe don't pass in null, so we don't have to make changes here?

File: src/main/default/lwc/lookup/lookup.js [hunk 18]
```diff
@@ -263,6 +267,17 @@ export default class Lookup extends LightningElement {
         this._hasFocus = false;
     }
 
+    handleOnClick(evt) {
+        if (!this.isSelectionAllowed()) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: let's look into this together, maybe we can find a way that does not change too much in the component and is not using attributes like focus that are not return the true focus status

**Review by itsmebasti — CHANGES_REQUESTED** [review]
there are to many non trivial changes, that will be hard to merge and maintain in the future, let's try to find a more "extending" solution.

File: src/main/default/lwc/lookup/lookup.js [hunk 19]
```diff
@@ -112,15 +119,16 @@ export default class Lookup extends LightningElement {
 
         // Compare clean new search term with current one and abort if identical
         const newCleanSearchTerm = newSearchTerm.trim().replace(/\*/g, '').toLowerCase();
-        if (this._cleanSearchTerm === newCleanSearchTerm) {
+
+        if (!this.allowEmptySearchQuery && this._cleanSearchTerm === newCleanSearchTerm) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: we aren't working with empty search terms anymore I thought?

File: src/main/default/lwc/lookup/lookup.js [hunk 20]
```diff
@@ -246,7 +254,7 @@ export default class Lookup extends LightningElement {
         this.template.querySelector('input').focus();
     }
 
-    handleFocus() {
+    handleFocus(evt) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: please revert all changes that do not have an effect on the changes we need.

File: src/main/default/lwc/lookup/lookup.js [hunk 21]
```diff
@@ -374,4 +391,8 @@ export default class Lookup extends LightningElement {
     get isExpanded() {
         return this.hasResults();
     }
+
+    get hasMoreResultsToDisplay() {
+        return this._hasFocus && this._searchResultsCount > this.resultsToDisplay;
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: can you explain how search result counts are related to hasFocus? maybe add a note if this is really needed

File: src/main/default/lwc/lookup/lookup.js [hunk 22]
```diff
@@ -48,13 +54,14 @@ export default class Lookup extends LightningElement {
...
     setSearchResults(results) {
+        this._searchResultsCount = results.length;
         // Reset the spinner
         this.loading = false;
         // Clone results before modifying them to avoid Locker restriction
         const resultsLocal = JSON.parse(JSON.stringify(results));
         // Format results
         const regex = new RegExp(`(${this._searchTerm})`, 'gi');
-        this._searchResults = resultsLocal.map((result) => {
+        this._searchResults = resultsLocal.splice(0, this.resultsToDisplay).map((result) => {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: do this in an extra line, so it is easier to merge and see.
Maybe also put the 2 custom lines at the same place, they are highly related to each other and this makes the custom changes more maintainable

File: src/main/default/lwc/lookup/lookup.js [hunk 23]
```diff
@@ -14,9 +14,14 @@ export default class Lookup extends LightningElement {
     @api isMultiEntry = false;
     @api hideResults = false;
     get showResults() {return this.isMultiEntry && !this.hideResults}
-    @api errors = [];
     @api scrollAfterNItems;
     @api iconResources = {};
+    @api errors = [];
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: move them back to ehre they were, to reduce the amount of changes we do to this class

File: src/main/default/lwc/lookup/lookup.html [hunk 24]
```diff
@@ -24,15 +25,15 @@
...
                             placeholder={placeholder}
                             value={getInputValue}
                             title={getInputTitle}
                             readonly={isInputReadonly}
                             onfocus={handleFocus}
                             onblur={handleBlur}
                             oninput={handleInput}
-                            onkeydown={handleKeyDown}
-                        />
+                            onkeydown={handleKeyDown} />
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.html]: revert

File: src/main/default/lwc/lookup/lookup.html [hunk 25]
```diff
@@ -146,13 +147,13 @@
             </template>
             <!-- Multi-selection end -->
 
-            <!-- Errors start -->
+            <!-- Error start -->
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.html]: revert all these changes

File: src/main/default/lwc/lookup/lookup.js [hunk 26]
```diff
@@ -1,7 +1,7 @@
 import { LightningElement, api } from 'lwc';
 
 const MINIMAL_SEARCH_TERM_LENGTH = 2; // Min number of chars required to search
-const SEARCH_DELAY = 300; // Wait 300 ms after user stops typing then, peform search
+const SEARCH_DELAY = 500; // Wait 300 ms after user stops typing then, perform search
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: revert these changes

File: src/main/default/lwc/lookup/lookup.js [hunk 27]
```diff
@@ -14,9 +14,14 @@ export default class Lookup extends LightningElement {
...
     @api hideResults = false;
     get showResults() {return this.isMultiEntry && !this.hideResults}
-    @api errors = [];
     @api scrollAfterNItems;
     @api iconResources = {};
+    @api errors = [];
+    @api disabled = false;
+    @api minimalSearchTermLength = MINIMAL_SEARCH_TERM_LENGTH;
+    @api resultsToDisplay = 100;
+    get minimalSearchQueryLength() {return this.allowEmptySearchQuery ? 0 : this.minimalSearchTermLength}
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: I thought we found another solution for empty searches?

File: src/main/default/lwc/lookup/lookup.js [hunk 28]
```diff
@@ -14,9 +14,14 @@ export default class Lookup extends LightningElement {
     @api isMultiEntry = false;
     @api hideResults = false;
     get showResults() {return this.isMultiEntry && !this.hideResults}
-    @api errors = [];
     @api scrollAfterNItems;
     @api iconResources = {};
+    @api errors = [];
+    @api disabled = false;
+    @api minimalSearchTermLength = MINIMAL_SEARCH_TERM_LENGTH;
+    @api resultsToDisplay = 100;
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: maxDisplayedResults

File: src/main/default/lwc/lookup/lookup.js [hunk 29]
```diff
@@ -33,6 +38,7 @@ export default class Lookup extends LightningElement {
     _defaultSearchResults = [];
     _curSelection = [];
     _focusedResultIndex = null;
+    _searchResultsCount = 0;
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js]: _numberOfAvailableSearchResults to make it more descriptive

File: src/main/default/lwc/lookup/lookup.js [hunk 30]
```diff
@@ -264,13 +272,21 @@ export default class Lookup extends LightningElement {
     }
 
     handleRemoveSelectedItem(event) {
+        if (this.disabled) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js:277]: maybe add a note, including the info that we can use the hooks soon

File: src/main/default/lwc/lookup/lookup.js [hunk 31]
```diff
@@ -264,13 +272,21 @@ export default class Lookup extends LightningElement {
...
+        }
+
         const recordId = event.currentTarget.name;
         this._curSelection = this._curSelection.filter((item) => item.id !== recordId);
         // Process selection update
         this.processSelectionUpdate(true);
     }
 
     handleClearSelection() {
+        if (this.disabled) {
```
Comment (itsmebasti) [inline src/main/default/lwc/lookup/lookup.js:288]: short question, what is the difference between the clear and the remove method?

File: src/main/default/classes/LookupSearchResult.cls [hunk 32]
```diff
@@ -17,6 +17,10 @@ public class LookupSearchResult implements Comparable {
         this.subtitle = subtitle;
     }
 
+    public LookupSearchResult(String id, String title, String icon) {
```
Comment (itsmebasti) [inline src/main/default/classes/LookupSearchResult.cls]: quick question: the sobject type is missing here, what is it used for, why can we skip it?

**Review by itsmebasti — CHANGES_REQUESTED** [review]
Let's reduce the custom changes to reduce merging effort

File: src/main/default/lwc/lookup/lookup.js (hunk 19 above)
Comment (rajas94) [reply src/main/default/lwc/lookup/lookup.js:123]: by default, we are not but if someone wants to work with an empty string, we allow it by using the `allowEmptySearchQuery` api property

File: src/main/default/lwc/lookup/lookup.js (hunk 21 above)
Comment (rajas94) [reply src/main/default/lwc/lookup/lookup.js]: we only want to display the helper text when the user is searching for something, as soon as he moves away from the input we don't want to display the text

File: src/main/default/lwc/lookup/lookup.js (hunk 31 above)
Comment (rajas94) [reply src/main/default/lwc/lookup/lookup.js:288]: `handleRemoveSelectedItem` is used to remove multi-select items and `handleClearSelection` is used to remove single entry lookup selection

File: src/main/default/classes/LookupSearchResult.cls (hunk 32 above)
Comment (rajas94) [reply src/main/default/classes/LookupSearchResult.cls:20]: I don't have all the fields in my use-case, that is why I have created another constructor
